StackOverflowMATLABchat/legtools
# Pull erikhuizinga/legtools/master into StackOverflowMATLABchat/legtools/master

# Summary

This is a large pull request fixing some bugs and adding a few new features, while remaining fully compatible with the previous release.
# Bugs fixed
## 1. `remidx` in `remove` method exceeds number of legend entries
### Example

``` matlab
figure
ezplot sin
legend sin
legtools.remove(legend,2)
% 2 exceeds the number of legend entries, which is 1
```

[This commit](https://github.com/StackOverflowMATLABchat/legtools/commit/9f55e4d5b39a9f4c33e6a039d99860263635a26c) fixes the bug.
## 2. `handlecheck` method improperly handles an empty legend object

`handlecheck` improperly throws an error when an empty legend object is passed to it. [This
commit](https://github.com/StackOverflowMATLABchat/legtools/commit/0cba18b8b4064617c3980d033c172ab989b5de32) implements detection of an empty legend.
## 3. `objtodelete` is never set in `remove` method
### Example

``` matlab
figure
fplot(@cos)
hold on
fplot(@sin)
legend cos sinlegtools.remove(legend,2)
```

Results in error:

```
Undefined function or variable 'objtodelete'.
Error in legtools.remove (line 134)
                delete(objtodelete);
```

[This commit](https://github.com/StackOverflowMATLABchat/legtools/commit/354a35e826bcbd5f6b3f7703e07c68f6250be834) fixes the bug.
## 4. `append` method improperly rearranges `Legend` property `PlotChildren`

The `append` method contains a bug when it appends to a legend that has been permuted previously or if the parent axes children have been sorted differently after the legend has been modified. `append` assumes the parent axes children are in the same order as the current legend `PlotChildren`, which may not be the case. For example, after permuting the legend entries, this order may have changed.
### Example

``` matlab
figure
fplot(@cos)
hold on
fplot(@sin)
legend cos sin
legtools.adddummy(legend,'dummy','ok')
legtools.permute(legend,[3 2 1])
legtools.remove(legend,2)
legtools.append(legend,'sin')
```

Which results in (notice the incorrect colouring of the legend entries):
![bug](https://cloud.githubusercontent.com/assets/19374736/19022914/d600a724-88e2-11e6-9b15-4ff4d6dedc84.png)

Note that the 3rd bug happens before this bug occurs, so you won't see this result on release v2.2. If you branch off [354a35e](https://github.com/StackOverflowMATLABchat/legtools/commit/354a35e826bcbd5f6b3f7703e07c68f6250be834) you can see the bug in action.

[This commit](https://github.com/StackOverflowMATLABchat/legtools/commit/ae66d0e06720bd3870c9d3c86a226621fab7ab7e) fixes the bug.
# Improvements
- Optimise code
  - Reduce code redundancy
  - Annotate with better comments
  - Improve readability
- Optimise help
  - Update help for new functionality
  - Format help to be easier to understand
  - Incorporate less technical explanations, hence easier for non-advanced MATLAB users
  - Format help to be more in line with help for MATLAB's built-in functions
  - Update Readme.md with latest help
  - Add new badge to Readme.md
  - Add new images to Readme.md
- Remove warning from `remove` method that was given when deleting the legend, because this is the expected behaviour (as explained in the help) and thus should not throw a warning
- Improve error text format
  - Make full sentences with capital letters and punctuation while being imperative like in most MATLAB functions
- Implement more input argument checks
- Improve detection of dummy legend entries
- Respect current hold state of parent axes of legend
# New functionality
- Support string data type introduced in MATLAB R2016b
- Enhance `plotParams` input argument of `adddummy` method
  - Support multiple formats
  - Make `plotParams` optional
